editor/src/components/canvas/ui-jsx-canvas.tsx

```diff
@@ -283,38 +283,47 @@ export const UiJsxCanvas = betterReactMemo(
 
     // TODO after merge requireFn can never be null
     if (requireFn != null) {
+      let resolvedFiles: string[] = []
       const customRequire = React.useCallback(
         (importOrigin: string, toImport: string) => {
           const filePathResolveResult = resolve(importOrigin, toImport)
           const resolvedParseSuccess: Either<string, MapLike<any>> = flatMapEither(
             (resolvedFilePath) => {
-              const projectFile = getContentsTreeFileFromString(projectContents, resolvedFilePath)
-              if (isTextFile(projectFile) && isParseSuccess(projectFile.fileContents.parsed)) {
-                const { scope } = createExecutionScope(
-                  resolvedFilePath,
-                  customRequire,
-                  mutableContextRef,
-                  topLevelComponentRendererComponents,
-                  projectContents,
-                  uiFilePath,
-                  transientFilesState,
-                  base64FileBlobs,
-                  hiddenInstances,
-                  metadataContext,
-                  shouldIncludeCanvasRootInTheSpy,
-                )
-                const exportsDetail = projectFile.fileContents.parsed.exportsDetail
-                let filteredScope: MapLike<any> = {}
-                for (const s of Object.keys(scope)) {
-                  if (s in exportsDetail.namedExports) {
-                    filteredScope[s] = scope[s]
-                  } else if (s === exportsDetail.defaultExport?.name) {
-                    filteredScope[s] = scope[s]
+              if (resolvedFiles.includes(resolvedFilePath)) {
+                // We're inside a cyclic dependency, so bail and the outer call will create the execution scope
+                return left('Ignoring inner cyclic dependency')
```

Comment:
This bit is a bit tricky. If you have the below set up:
`a.js`:
```
import { B } from './b'

export const C = (props) => {
  return <div/>
}
export const A = (props) => {
  return <B/>
}
```
`b.js`:
```
import { C } from './a'

export const B = (props) => {
  return <C/>
}
```

Then you'll be able to focus `A/B`, but not `A/B/C` (and so won't be able to edit the final underlying `div`).

I tried experimenting with retaining a `resolvedScopes` instead, like so:
```
let resolvedScopes: { [filepath: string]: any } = {}
...
if (resolvedFilePath in resolvedScopes) {
  return right(resolvedScopes[resolvedFilePath])
} else {
  let filteredScope: MapLike<any> = {}
  resolvedScopes[resolvedFilePath] = filteredScope
```
(where `filteredScope` is then filled in below as before), but that results in `C` being `undefined` when rendering.

Reply:
this is so weird. but also this is pretty much an edge case so I will resist the temptation to fall into this rabbit hole

Reply:
I'm going to create a follow up ticket on this and add it to the backlog - this solution works because returning the `left` means we fall back to the regular require function, meaning that (as mentioned above) a component that is in a dependency cycle won't be spy wrapped (though one in a file with a cyclic dependency will), so we still are left with the above edge case, but that doesn't seem as high priority as fixing the far more common case (which this does).